File: Categories/Union_Find/1101_theEarliestMomentWhenEveryoneBecomeFriends.cs
public class UnionFind {
    int[] parents;
    int[] ranks;

    public UnionFind(int n) {
        parents = new int[n];
        for (int idx = 0; idx < n; idx++) {
            parents[idx] = idx;
        }
        
        ranks = new int[n];
        for (int idx = 0; idx < n; idx++) {
            ranks[idx] = 1;
        }
    }

    public int find(int n) {
        int root = n;
        while (root != parents[root]) {
            root = parents[root];
        }
        
        int cur = n;
        int nxt;
        while (cur != root) {
            nxt = parents[cur];
            parents[cur] = root;
            cur = nxt;
        }

        return root;
    }

    public bool union(int n1, int n2) {
        int p1 = find(n1), p2 = find(n2);
        if (p1 == p2) {
            return false; 
        }

        if (ranks[p1] > ranks[p2]) {
            ranks[p2] += ranks[p1];
            parents[p1] = p2;
        }
        else {
            ranks[p1] += ranks[p2];
            parents[p2] = p1;
        }

        return true;
    }
}

public class Solution {
    public int EarliestAcq(int[][] logs, int n) {
        logs = logs.OrderBy(log => log[0]).ToArray();
        UnionFind union_find = new UnionFind(n);

        int t, n1, n2;
        int num_groups = n;
        for (int idx = 0; idx < logs.Length; idx++) {
            t = logs[idx][0];
            n1 = logs[idx][1];
            n2 = logs[idx][2];

            if (union_find.union(n1, n2)) {
                num_groups -= 1;
            }

            if (num_groups == 1) {
                return t;
            }
        }

        return -1;
    }
}

File: Categories/Union_Find/1101_theEarliestMomentWhenEveryoneBecomeFriends.py
from typing import *


class UnionFind:
    def __init__(self, n: int) -> None:
        self.parents = [idx for idx in range(n)]
        self.ranks = [1 for _ in range(n)]
    
    def find(self, n: int) -> int:
        root = n
        while root != self.parents[root]:
            root = self.parents[root]
        
        cur = n
        while cur != root:
            nxt = self.parents[cur]
            self.parents[cur] = root
            cur = nxt
        
        return root
    
    def union(self, n1: int, n2: int) -> bool:
        p1, p2 = self.find(n1), self.find(n2)
        if p1 == p2:
            return False
        
        if self.ranks[p1] > self.ranks[p2]:
            self.ranks[p1] += self.ranks[p2]
            self.parents[p2] = p1
        else: 
            self.ranks[p2] += self.ranks[p1]
            self.parents[p1] = p2
        
        return True

class Solution:
    def earliestAcq(self, logs: List[List[int]], n: int) -> int:
        logs.sort()

        union_find = UnionFind(n)

        num_groups = n
        for log in logs:
            t, n1, n2 = log
            if union_find.union(n1, n2):
                num_groups -= 1
        
            if num_groups == 1:
                return t
        
        return -1